java/java.lsp.server/src/org/netbeans/modules/java/lsp/server/protocol/LspTemplateUI.java

```diff
@@ -0,0 +1,215 @@
+                JsonPrimitive jp = (JsonPrimitive) arg;
+                path = jp.getAsString();
+            } else {
+                path = arg.toString();
+            }
+            File file = new File(path);
+            if (file.exists()) {
+                target[0] = DataFolder.findFolder(FileUtil.toFileObject(file));
+            }
+        }
+        return target;
+    }
+
+    private static String removeExtensionFromFileName(String nameWithExtension, String templateExtension) {
+        if (nameWithExtension.endsWith('.' + templateExtension)) {
+            return nameWithExtension.substring(0, nameWithExtension.length() - templateExtension.length() - 1);
+        } else {
+            return nameWithExtension;
+        }
+    }
+
+    private static List<QuickPickItem> quickPickTemplates(final DataFolder folder) {
+        Node[] arr = folder.getNodeDelegate().getChildren().getNodes(true);
+        List<QuickPickItem> categories = new ArrayList<>();
+        for (Node n : arr) {
+            FileObject fo = n.getLookup().lookup(FileObject.class);
+            DataObject obj = n.getLookup().lookup(DataObject.class);
+
+            boolean display;
+            if (obj instanceof DataFolder) {
+                Object o = obj.getPrimaryFile ().getAttribute ("simple"); // NOI18N
+                display = o == null || Boolean.TRUE.equals (o);
+            } else {
+                display = obj.isTemplate();
+            }
+
+            if (display) {
+                String detail = findDetail(obj);
+                final String displayName = n.getDisplayName();
+                String description = n.getShortDescription();
+                if (description != null && description.equals(displayName)) {
+                    description = null;
+                }
+                categories.add(new QuickPickItem(
+                    displayName, description, detail,
+                    false, fo.getNameExt()
+                ));
+            }
+        }
+        return categories;
+    }
+
+    private static String findDetail(DataObject obj) {
+        URL description = org.openide.loaders.TemplateWizard.getDescription(obj);
+        String descriptionText = null;
+        if (description != null) {
+            try (ByteArrayOutputStream os = new ByteArrayOutputStream(); InputStream is = description.openStream()) {
+                FileUtil.copy(is, os);
+                descriptionText = os.toString("UTF-8");
+                int bodyBegin = descriptionText.toUpperCase().indexOf("<BODY>");
+                int bodyEnd = descriptionText.toUpperCase().indexOf("</BODY>");
+                if (bodyBegin >= 0 && bodyEnd > bodyBegin) {
+                    descriptionText = descriptionText.substring(bodyBegin + 6, bodyEnd);
+                } else {
+                    descriptionText = null;
+                }
+            } catch (IOException ex) {
+                throw raise(RuntimeException.class, ex);
```

Comment:
Maybe not abort the entire operation when the detail / description can't be fetched, just log.

Reply:
Very true. I had to fix it in d9b0c5c as the HTML/JavaScript category couldn't be selected otherwise.